src/terms.ml

```diff
@@ -117,6 +119,41 @@ type term =
   ; vars  : (string * int) array
   (** Name and arity of the pattern variables bound in the RHS. *) }
 
+(** Trees are used to efficiently choose a rewriting rule given a list of
+    terms (beginning with a symbol) to be rewritten.  The left hand has side
+    of the rule--being the pattern to be matched for the rule to apply--is
+    spread across the {!cons:Node}s of the tree.  When a {!cons:Leaf} is
+    reached, the target is rewrote to the content of the leaf. *)
+ and tree =
+    Leaf of (term_env, term) Bindlib.mbinder
+  (** Hold the targets of rewriting, right hand side of a rule.  In a
+      {!cons:Leaf}[(t, a)], [t] contains the term--if a switch has been
+      performed-- which has been matched to arrive to the leaf or {!cons:None}
+      if we arrived on the leaf via a default case. [a] is the right hand side
+      of the rule, or the action to perform if the rule applies.
+      XXX Remove term option from doc *)
+  | Node of node_data
+  (** Nodes allow to perform switches, a switch being the matching of a
+      pattern.  Briefly, a {!cons:Node} contains one subtree per possible
+      switch, plus possibly a default case. *)
+  | Fail
+
+(** Data contained in a node of the tree.  {!recfield:switch} contains the
+    term on which the switch that gave birth to this node has been performed.
+    {!recfield:swap} indicates whether the columns of the matrix have been
+    swapped before the switch and {!recfield:children} contains the
+    subtrees. *)
+ and node_data =
+  { swap : int option
+  (** Indicates which column of the pattern matrix has been chosen to perform
+      the switch.
+      XXX remove the option *)
+  ; children : (term option * tree) list
+  (** Subtrees resulting from either specialisation on terms or default case.
+      First element is {!cons:None} if child is result of a default case or
+      {!cons:Some}[(t)] if it results from specialisation on [t]. *) }
+(* TODO remove switch field, redundant with new [children] type *)
+
```

Comment:
I don't understand what is this term option. Shouldn't it be just a symbol option? Why do you use the type term?

Reply:
A term is used as we can specialise on an abstraction as well, can't we?  This 'issue' has already been mentioned some time ago, and you suggested to create a new datatype

> Why not using a type Const of symb | Lambda?

Which I'd like to do, but I'd have to adapt the prettyprinter functions